File: binarySearch/search_2D_matrix12.java
//solution1: sorted in one direction
public boolean searchMatrix(int[][] matrix, int target){
	int[] res = new int[]{-1,-1};
	if(matrix == null || matrix.length == 0)
		return res;
	if(matrix[0] == null || matrix[0].length == 0)
		return res;

	int row = matrix.length;
	int col = matrix[0].length;
	int left = 0;
	int right = row*col -1; //go through all element in matrix

	while(left <= right){
		int mid = left + (right - left)/2;
		//given index of array in matrix, get position in matrix
		int midx = mid/col;
		int midy = mid%col;
		int value = matrix[midx][midy];
		if(value == target)
			return new int[]{midx, midy};
		else if(value > target)
			right = mid-1;
		else
			left = mid+1;
	}
	return res;
}
//sorect in x and sorted in y
https://leetcode.com/problems/search-a-2d-matrix-ii/discuss/66140/My-concise-O(m%2Bn)-Java-solution
//do binary search in 2 directions  O(m+n)
public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length < 1 || matrix[0].length <1) {
            return false;
        }
        int col = matrix[0].length-1;
        int row = 0;
        while(col >= 0 && row <= matrix.length-1) {
            if(target == matrix[row][col]) {
                return true;
            } else if(target < matrix[row][col]) {
                col--;
            } else if(target > matrix[row][col]) {
                row++;
            }
        }
        return false;
    }

File: binarySearch/search_2D_matrix12.py
# sorted in one direction
'''
matrix = [
  [1,   3,  5,  7],
  [10, 11, 16, 20],
  [23, 30, 34, 50]
]
'''
def search2DMatrix1(matrix, target):
	if matrix is None or not matrix:
		return False

	row = len(matrix)
	col = len(matrix[0])
	left = 0
	right = row * col -1

	while left <= right
		mid = left + (right - left) // 2
		x = mid // row
		y = mid * row
		if matrix[x][y] == target:
			return True
		elif matrix[x][y] < target:
			left = mid + 1
		else:
			right = mid - 1
	return False

# sorted in 2 direction, both x and y
'''
[
  [1,   4,  7, 11, 15],
  [2,   5,  8, 12, 19],
  [3,   6,  9, 16, 22],
  [10, 13, 14, 17, 24],
  [18, 21, 23, 26, 30]
]
'''
# https://leetcode.com/problems/search-a-2d-matrix-ii/discuss/66140/My-concise-O(m%2Bn)-Java-solution
# do binary search in 2 directions  O(m+n)
def search2DMatrix2(matrix, target):
	if matrix is None or not matrix:
		return False
	# Start from the bottom left most corner, or top right corner
	row = 0
	col = len(matrix[0]) - 1

	while row < len(matrix) and col >= 0
		if matrix[row][col] == target:
			return True
		elif matrix[row][col] < target:
			row += 1
		else:
			col -= 1
	return False
